### ISSUE TITLE
mapData doesn't set color correcty if color range is the same

### ISSUE THREAD

**Issue by vrynkov** [cytoscape/cytoscape.js#2152]
**Issue type**

Bug report

**Environment info**

- Cytoscape.js version : 3.2.11
- Browser/Node.js & version : Chrome 67.0 (not browser-specific)

**Current (buggy) behaviour**

I build CSS for Cytoscape.js.  
My selector is 
selector='edge[score > 2.5002426779747005][score < 4.4712]'

My mapData is
line-color='mapData(score,2.5002426779747005,4.4712,rgb(255,255,0),rgb(255,255,0))'

I expect the line color to be (255,255,0) for the range value of score
2.5002426779747005 ... 4.4712  

The line color is incorrectly set to gray. 

**Desired behaviour**

The line color for edge whose value ranges between 2.5002426779747005 ... 4.4712  
should be (255,255,0) - yellow.

**Minimum steps to reproduce**

**Comment by maxkfranz** [comment]
The mapper isn't technically valid, since the range of the output values is zero.  An example of a valid mapper would be:

```
edge[score > 2.5002426779747005][score < 4.4712] {
  line-color: rgb(255,255,0);
}
```

Your mapper is also redundant based on the selector, which would result in unnecessary overhead.  The effect would become more pronounced as the size of the stylesheet grows.

The mapping routine could fail more gracefully in cases like these by using the single value, but it should also then at least print a warning message to indicate the error.  I don't see a case for creating a mapping over a range of zero width rather than just using a single value.

**Comment by vrynkov** [comment]
Yes, I modified our code to check the range of values mapped to in  mapData; if they are the same (the last two arguments), then I just use the constant (like you showed in your example above).

**Comment by vrynkov** [comment]
I just wanted to let you know about this case.  It still seems to me that mapData should set the color to (255,255,0) and not to a random (gray) color.  But that is fine, we can close this issue if it is not going to be changed.  Thank you for looking into it.

**Comment by maxkfranz** [comment]
Invalid mappings like these are logged to the console in 3.2:

Ref : https://github.com/cytoscape/cytoscape.js/blob/v3.2.14/src/style/parse.js#L138

E.g.:

> The style property `background-color: mapData(weight, 0, 100, red, red)` is invalid

Cytoscape logs messages like these to the browser's console to indicate things that the client should fix in his code.  This is very similar to how the browser handles invalid CSS.  Also like CSS in the browser, the library will throw away invalid definitions from the style block --- so the specified value in the closest previous block is used in lieu of the invalid one.

In 3.3, Cytoscape will autocorrect invalid definitions of this type with a warning message.  Generally though, erronous definitions ought to be corrected before being passed.

See : Autocorrect invalid `mapData()` style property mappers that have zero output range #2153
